dropwizard-jdbi/src/main/java/io/dropwizard/jdbi/DBIFactory.java

```diff
@@ -83,33 +88,104 @@ private SanerNamingStrategy() {
+     * with the configured health check; this method should not be overridden
+     * (instead, override {@link #configure(DBI, String)} and/or
+     * {@link #newInstance(ManagedDataSource)})
+     * 
+     * @param environment
+     * @param configuration
+     * @param dataSource
+     * @param name
+     * @return A fully configured {@link DBI} object
+     */
     public DBI build(Environment environment,
                      PooledDataSourceFactory configuration,
                      ManagedDataSource dataSource,
                      String name) {
-        final String validationQuery = configuration.getValidationQuery();
-        final DBI dbi = new DBI(dataSource);
+
+        // Create the instance
+        final DBI dbi = this.newInstance(dataSource);
+        
+        // Manage the data source that created this instance.
         environment.lifecycle().manage(dataSource);
+        
+        // Setup the required health checks.
+        final String validationQuery = configuration.getValidationQuery();
         environment.healthChecks().register(name, new DBIHealthCheck(
                 environment.getHealthCheckExecutorService(),
                 configuration.getValidationQueryTimeout().orElseGet(() -> Duration.seconds(5)),
                 dbi,
                 validationQuery));
+
+        // Setup logging.
         dbi.setSQLLog(new LogbackLog(LOGGER, Level.TRACE));
+        
+        // Setup the timing collector
         dbi.setTimingCollector(new InstrumentedTimingCollector(environment.metrics(),
                 new SanerNamingStrategy()));
+        
         if (configuration.isAutoCommentsEnabled()) {
             dbi.setStatementRewriter(new NamePrependingStatementRewriter(new ColonPrefixNamedParamStatementRewriter()));
         }
-        dbi.registerArgumentFactory(new GuavaOptionalArgumentFactory(configuration.getDriverClass()));
-        dbi.registerArgumentFactory(new OptionalArgumentFactory(configuration.getDriverClass()));
+
+        // Add the default argument and column mapper factories.
+        this.configure(dbi, configuration);
+
+        return dbi;
+    }
+    
+    /**
+     * This creates a vanilla DBI instance based on the specified data source;
+     * this can be overridden if required
+     * 
+     * @param dataSource
+     * @return
+     */
+    protected DBI newInstance(final ManagedDataSource dataSource) {
+        return new DBI(dataSource);
+    }
+
+    /**
+     * This method sets the default column mapper and argument factories
+     * <p>
+     * if this is overridden it is strongly recommend that
+     * {@code super.configure(dbi, configuration)} is invoked before any other
+     * changes are made if you intend to use the default as a base so that the
+     * customized argument settings will supersede the defaults
+     * </p>
+     * 
+     * @param dbi
+     * @param configuration
+     */
+    protected void configure(final DBI dbi, final PooledDataSourceFactory configuration) {
```

Comment:
There could be another approach. We could create a minimally defined interface that is parent of PooledDataSourceFactory that has only the method(s) we need for setting up the default argument settings. Then, the configure method could take this interface as a second argument. If this interface has only one method, say `getDriverClass()`, then for easy testing you could do something like this:

```Java
configure(new DBI("url"), () -> "io.dropwizard.driver.Driver");
```